File: zen/vm.go
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"unsafe"

	"github.com/ngaut/log"
)

type ValueType int

const (
	TypeUnit ValueType = iota
	TypeInt
	TypeBool
	TypeTuple
	TypeClosure
	TypeString
	TypeReturnAddr
	TypeEnv
)

type Value struct {
	ValueType
}

type Unit struct {
	Value
}

type Int struct {
	Value
	v int
}

type ReturnAddr struct {
	Value
	v int
}

func NewReturnAddr(pc int) *Value {
	tmp := ReturnAddr{v: pc}
	tmp.ValueType = TypeReturnAddr
	return &tmp.Value
}

type Bool struct {
	Value
	v bool
}

type Tuple struct {
	Value
	tag   uint8
	tuple []*Value
}

type Closure struct {
	Value
	env  *Env
	code []byte
	pc   int
}

func (v *Value) Type() ValueType {
	return v.ValueType
}

func NewInt(v int) *Value {
	tmp := Int{v: v}
	tmp.Value.ValueType = TypeInt
	return &tmp.Value
}

func NewBool(v bool) *Value {
	tmp := Bool{v: v}
	tmp.Value.ValueType = TypeBool
	return &tmp.Value
}

func NewTuple(tag uint8, tuple []*Value) *Value {
	tmp := Tuple{tag: tag, tuple: tuple}
	tmp.ValueType = TypeTuple
	return &tmp.Value
}

func NewClosure(pc int, env *Env) *Value {
	tmp := Closure{env: env, pc: pc}
	tmp.ValueType = TypeClosure
	return &tmp.Value
}

func NewEnv(data []*Value) *Value {
	tmp := Env{}
	tmp.data = make([]*Value, len(data))
	copy(tmp.data, data)
	tmp.ValueType = TypeEnv
	return &tmp.Value
}

func (v *Value) Int() *Int {
	return (*Int)(unsafe.Pointer(v))
}

func (v *Value) Bool() *Bool {
	return (*Bool)(unsafe.Pointer(v))
}

func (v *Value) Closure() *Closure {
	return (*Closure)(unsafe.Pointer(v))
}

func (v *Value) Env() *Env {
	return (*Env)(unsafe.Pointer(v))
}

func (v *Value) Tuple() *Tuple {
	return (*Tuple)(unsafe.Pointer(v))
}

func (v *Value) ReturnAddr() *ReturnAddr {
	return (*ReturnAddr)(unsafe.Pointer(v))
}

type Env struct {
	Value
	data []*Value
}

func (e *Env) Get(n int) *Value {
	return e.data[n]
}

type Stack struct {
	data []*Value
	top  int
}

func (s *Stack) PopN(n int) []*Value {
	ret := s.data[s.top-n : s.top]
	s.top -= n
	return ret
}

func (s *Stack) Push(v *Value) {
	s.data[s.top] = v
	s.top++
}

func (s *Stack) Access(n int) *Value {
	return s.data[s.top-n]
}

type ctx struct {
	mark  int
	stack []int
	sp    int
}

func (c *ctx) Push(v int) {
	c.stack[c.sp] = v
	c.sp++
}

func (c *ctx) Pop() int {
	c.sp--
	return c.stack[c.sp]
}

type VM struct {
	pc    int
	sp    int
	bp    int
	acc   *Value
	env   *Env
	stack []*Value

	ctx
}

func New() *VM {
	vm := &VM{
		stack: make([]*Value, 200),
		env:   &Env{},
	}
	vm.ctx.stack = make([]int, 100)
	return vm
}

func (vm *VM) Run(code []byte) *Value {
LOOP:
	for {
		switch code[vm.pc] {
		case STOP:
			log.Debug("STOP")
			break LOOP
		case MARK:
			log.Debugf("MARK sp=%d", vm.sp)
			vm.ctx.mark = vm.sp
			vm.pc++
		case CONST:
			v := binary.LittleEndian.Uint32(code[vm.pc+1:])
			log.Debug("CONST", v)
			vm.acc = NewInt(int(v))
			vm.pc += 5
		case ConstTrue:
			vm.acc = NewBool(true)
			vm.pc++
		case ConstFalse:
			vm.acc = NewBool(false)
			vm.pc++
		case MakeTuple:
			size := int(code[vm.pc+2])
			tuple := make([]*Value, size)
			copy(tuple, vm.stack[vm.sp-size:])
			vm.acc = NewTuple(code[vm.pc+1], tuple)
			vm.pc += 3
		case CLOSURE:
			size := binary.LittleEndian.Uint32(code[vm.pc+1:])
			log.Debugf("CLOSURE %d", size)
			vm.acc = NewClosure(vm.pc+5, vm.env.Env())
			vm.pc += int(size + 5)
		case APPLY:
			log.Debug("APPLY")
			cls := vm.acc.Closure()
			// 保存之前的pc
			vm.ctx.Push(vm.pc + 1)
			vm.pc = cls.pc
		case CHECK:
			n := code[vm.pc+1]
			if vm.sp-vm.ctx.mark >= int(n) {
				log.Debugf("CHECK, want args %d, have args %d", n, vm.sp-vm.ctx.mark)
				vm.pc += 2
			} else {
				log.Debug("CHECK partial apply")
				// TODO handle partial apply
				panic("not implement")
			}
		case ENV:
			// copy env to stack
			env := vm.acc.Closure().env
			for _, v := range env.data {
				vm.stack[vm.sp] = v
				vm.sp++
			}
			log.Debugf("ENV, copy %d stack, save bp %d, set bp to sp %d", len(env.data), vm.bp, vm.sp)
			vm.ctx.Push(vm.bp) // 保存bp
			vm.bp = vm.sp
			vm.pc++
		case UNENV:
			vm.bp = vm.ctx.Pop()
			log.Debugf("UNENV, recover bp %d", vm.bp)
			vm.pc++
		// case PUSHADDR:
		// 	size := binary.LittleEndian.Uint32(code[vm.pc+1:])
		// 	vm.stack[vm.sp] = NewReturnAddr(vm.pc + int(size))
		// 	vm.sp++
		// 	vm.stack[vm.sp] = &vm.env.Value
		// 	vm.sp++
		// 	vm.pc += 4
		case ACCESS:
			n := int(code[vm.pc+1])
			vm.stack[vm.sp] = vm.stack[vm.bp-n]
			vm.sp++
			vm.pc += 2
			log.Debugf("ACCESS %d", n)
		case PUSH:
			log.Debug("PUSH")
			vm.stack[vm.sp] = vm.acc
			vm.sp++
			vm.pc++
		case ADDINT:
			log.Debug("ADDINT")
			top := vm.stack[vm.sp-1].Int()
			vm.acc.Int().v += top.v
			vm.sp--
			vm.pc++
		case RETURN:
			vm.sp = vm.bp
			vm.pc = vm.ctx.Pop()
			log.Debugf("RETURN sp=%d", vm.sp)
		case JumpIf: // + int
		case Goto: // + int
		case Field: // + int
		case Switch: // + (tag + offset) * n
		default:
			panic("not implement")
		}
	}
	return vm.acc
}

type Emit struct {
	bytes.Buffer
}

func (e *Emit) CONST(v int32) {
	e.WriteByte(CONST)
	binary.Write(&e.Buffer, binary.LittleEndian, v)
}

func (e *Emit) PUSH() {
	e.WriteByte(PUSH)
}

func (e *Emit) CLOSURE(c []byte) {
	e.WriteByte(CLOSURE)
	binary.Write(&e.Buffer, binary.LittleEndian, int32(len(c)))
	e.Write(c)
}

func (e *Emit) CHECK(n byte) {
	e.WriteByte(CHECK)
	e.WriteByte(n)
}

func (e *Emit) ACCESS(n byte) {
	e.WriteByte(ACCESS)
	e.WriteByte(n)
}

func (e *Emit) ADDINT() {
	e.WriteByte(ADDINT)
}

func (e *Emit) RETURN() {
	e.WriteByte(RETURN)
}

func (e *Emit) APPLY() {
	e.WriteByte(APPLY)
}

func (e *Emit) STOP() {
	e.WriteByte(STOP)
}

func (e *Emit) MARK() {
	e.WriteByte(MARK)
}

func (e *Emit) ENV() {
	e.WriteByte(ENV)
}

func (e *Emit) UNENV() {
	e.WriteByte(UNENV)
}

const (
	STOP = iota
	MARK
	CONST
	PUSH
	CLOSURE
	APPLY
	CHECK
	ENV
	ACCESS
	ADDINT
	UNENV
	RETURN
	BRANCH

	ConstTrue
	ConstFalse
	MakeTuple // + tag + size
	TailApply
	JumpIf // + int
	Goto   // + int
	Field  // + int
	Switch // + (tag + offset) * n
)

func main() {
	c := &Emit{}
	c.CHECK(2)
	c.ENV()
	c.ACCESS(0)
	c.PUSH()
	c.ACCESS(1)
	c.ADDINT()
	c.UNENV()
	c.RETURN()

	d := &Emit{}
	d.MARK()
	d.CONST(1)
	d.PUSH()
	d.CONST(2)
	d.PUSH()
	d.CLOSURE(c.Bytes())
	d.APPLY()
	d.STOP()

	fmt.Println("hello wolrd:", d.Bytes())
	vm := New()
	vm.Run(d.Bytes())
	fmt.Println(vm.acc)
}

File: zen/vm.ml
let env_get (e : result list) n = List.nth e n

let env_put (e : result list) v = v :: e


type result =
    Value of int
  | Bool of bool
  | Tuple of int * result list
  | Prim of string
  | Lambda of (Instruct.t list * result list)
  | Eplison

let step (c, e, s, r) op =
  match op with
  Instruct.Bool v -> Stack.push (Bool v) s; (c, e, s, r)
  | Instruct.Const v -> Stack.push (Value v) s; (c, e, s, r)
  | Instruct.Prim str -> Stack.push (Prim str) s; (c, e, s, r)
  | Instruct.MakeTuple (tag, n) ->
    let rec loop i n res =
      if i=n then begin Stack.push (Tuple (tag, res)) s; (c, e, s, r) end
      else loop (i+1) n ((Stack.pop s)::res)
    in loop 0 n []
  | Instruct.Field n ->
    (match (Stack.pop s) with
    | Tuple (tag, ls) -> (Stack.push (List.nth ls n) s); (c, e, s, r)
    | _ -> failwith "type infer should kill this case")
  | Instruct.Pop -> Stack.pop s |> ignore; (c, e, s, r)
  | Instruct.Copy -> Stack.push (Stack.top s) s |> ignore; (c, e, s, r)
  | Instruct.Plus -> (match Stack.pop s with
    | Value x -> (match Stack.pop s with
        | Value y -> Stack.push (Value (x+y)) s;
          (c, e, s, r)
        | _ -> failwith "can add non int")
    | _ -> failwith "can add non int")
  | Instruct.Sub -> (match Stack.pop s with
    | Value x -> (match Stack.pop s with
        | Value y -> Stack.push (Value (y-x)) s;
          (c, e, s, r)
        | _ -> failwith "can add non int")
    | _ -> failwith "can add non int")
  | Instruct.Mul -> (match Stack.pop s with
    | Value x -> (match Stack.pop s with
        | Value y -> Stack.push (Value (x*y)) s;
          (c, e, s, r)
        | _ -> failwith "can add non int")
    | _ -> failwith "an add non int")
  | Instruct.Equal -> (match Stack.pop s with
    | Value x -> (match Stack.pop s with
        | Value y -> Stack.push (Bool (if x=y then true else false)) s;
          (c, e, s, r)
        | _ -> failwith "can add non int")
    | _ -> failwith "can add non int")
  | Instruct.Branch (succ, fail) ->
    (match Stack.pop s with
     | Bool x -> if x then
         (succ, e, s, r)
       else (fail, e, s, r)
     | _ -> failwith "branch instruct expect bool type")
  | Instruct.Access n -> Stack.push (env_get e n) s; (c, e, s, r)
  | Instruct.Closure c1 -> Stack.push (Lambda (c1,e)) s; (c, e, s, r)
  | Instruct.Tailapply ->
    let (c1,e1) = match Stack.top s with
        Lambda x -> x
      | _ -> failwith "cannot tailapply with non closure" in
    (c1, e1, s, r)
  | Instruct.Apply ->
    (match Stack.top s with
     | Lambda (c1, e1) ->
       Stack.push (Lambda (c,e)) r;
       (c1, e1, s, r)
     | _ -> failwith "cannot apply non closure")
  | Instruct.Pushmark -> Stack.push Eplison s; (c, e, s, r)
  | Instruct.Grab ->
    if Stack.top s = Eplison then
      (Stack.pop s |> ignore;
       Stack.push (Lambda (Instruct.Grab::c,e)) s;
       (let (c1,e1) = match Stack.pop r with
             Lambda x -> x
           | _ -> failwith "should be closure in r" in
        (c1, e1, s, r)))
    else let v = Stack.pop s in
      (c, (env_put e v), s, r)
  | Instruct.Return ->
    let v = Stack.pop s in
    (if Stack.top s = Eplison then
      (Stack.pop s |> ignore; Stack.push v s;
       let (c1,e1) = match Stack.pop r with
           Lambda x -> x
         | _ -> failwith "should be closure in return" in
       (c1, e1, s, r))
    else match v with
        Lambda (c1,e1) -> (c1, e1, s, r)
      | _ -> failwith "should be")
  | Instruct.Switch cases ->
    (match Stack.pop s with
    | Tuple (tag,_) ->
      let code = List.assq tag cases in
      (code, e, s, r)
    | _ -> failwith "switch should get tuple!")
  | _ -> failwith "not implement"

let run code e s r =
  let rec loop (c, e, s, r) =
    match c with
    | [] -> failwith "must stop with Instruct.Stop"
    | [Instruct.Stop] -> Stack.top s
    | op::c1 -> loop (step (c1, e, s, r) op)
  in loop (code, e, s, r)
